### PR TITLE
Fixed bug where querystring was lost on connection upgrade without using request location

### LINKED ISSUE #89360: Kubectl proxy would strip out the parameter for websocket request
**What happened**:
Create a deployment to run a websocket server. Use `kubectl proxy` for accessing the service outside of the cluster.

**What you expected to happen**:

Invoke the URL, for example as below, would pas in the token to the websocket server.
```
"ws://..../api/v1/namespaces/..../services/theservice/proxy/websocket?token=123"
```

However, instead, the parameter is stripped out.

image

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.17.1
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Darwin
- Kernel (e.g. `uname -a`): Darwin Kernel Version 19.2.0:
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by brianpursley**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fixes a bug where websocket requests going through kubectl proxy were losing their querystring parameters.

**Which issue(s) this PR fixes**:
Fixes #89360 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

### CODE REVIEW

**Review by tarokkk — APPROVED** [review]

**Comment by brianpursley** [comment]
I think this PR fixes the bug. I'm OK with closing this if change is not wanted, I'll understand. But I don't want to let it go stale.

/remove-lifecycle stale

**Comment by Jefftree** [comment]
Hi, so sorry I missed this earlier.

/lgtm

/assign @caesarxuchao

**Comment by kwiesmueller** [comment]
This still sounds like a good change for me (and a problem I ran into recently).
@Jefftree @caesarxuchao could you have a look.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 1]
```diff
@@ -265,6 +265,11 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 		location = *req.URL
 		location.Scheme = h.Location.Scheme
 		location.Host = h.Location.Host
+	} else if len(location.RawQuery) == 0 {
```
Comment (kwiesmueller) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:268]: Shouldn't we combine the queries if they exist on both?
Not sure I understand when this would happen and we'd still drop the query and if this is expected by the user.

Comment (brianpursley) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: Good call.

At first I thought this might be too difficult and settled on this approach, but on second look it seems pretty simple to handle.  I updated the PR to merge the location and request queries.  ptal.

Comment (kwiesmueller) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: Thanks!
/lgtm

**Comment by Jefftree** [comment]
The tests failures look legitimate and seem to be tied to this PR after combining the queries. I'm not too familiar with the tests, but it'd probably be worth investigating if merging is the desired behavior.

```
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:388
Mar 11 18:01:48.961: Unexpected kubectl exec output. Wanted "running in container", got "running in container echo running in container"
```

**Comment by brianpursley** [comment]
@Jefftree yeah, I definitely broke those tests, they fail when I run them locally too.  I'll take a look at it again and see if I can figure out what is going on.  Its been a long time since I originally worked on this, so I will need to regain my context and how to reproduce the issue so I can properly test the fix.

**Comment by brianpursley** [comment]
I'm still looking into this, but I made a comment on the original issue with how I am able to reproduce the issue locally 
https://github.com/kubernetes/kubernetes/issues/89360#issuecomment-797207790

**Comment by brianpursley** [comment]
I'm going back to what I originally had :sweat_smile:  .  I think it is the only safe way to handle it.  

Merging the query strings worked for proxy, but broke exec, which broke a bunch of e2e tests as you saw.

Running the tests locally, I can see the problem seems to be that it is duplicating the parameters for exec.   Example below from kubelet.log when I ran one of the failing e2e tests:

Current release version of code (with none of my changes). This is when the test passed. You can see exec has `command=cat&command=/etc/hosts/`
```
I0311 22:41:36.516313  522022 httplog.go:89] "HTTP" verb="POST" URI="/exec/e2e-kubelet-etc-hosts-794/test-pod/busybox-1?command=cat&command=%2Fetc%2Fhosts&error=1&output=1" latency="97.498713ms" userAgent="e2e.test/v1.21.0 (linux/amd64) kubernetes/8a1f30a -- [sig-node] KubeletManagedEtcHosts should test kubelet managed /etc/hosts file [LinuxOnly] [NodeConformance] [Conformance]" srcIP="127.0.0.1:50428" hijacked=true
```

After my previous attempt to merge the location and request query strings, it duplicated the commands and added a whole bunch more keys that exec doesn't want. This is when the same e2e test failed:
```
I0311 22:32:41.988978  504654 httplog.go:89] "HTTP" verb="POST" URI="/exec/e2e-kubelet-etc-hosts-8088/test-pod/busybox-1?command=cat&command=%2Fetc%2Fhosts&command=cat&command=%2Fetc%2Fhosts&container=busybox-1&container=busybox-1&error=1&output=1&stderr=true&stdout=true" latency="102.314984ms" userAgent="e2e.test/v1.21.0 (linux/amd64) kubernetes/8a1f30a -- [sig-node] KubeletManagedEtcHosts should test kubelet managed /etc/hosts file [LinuxOnly] [NodeConformance] [Conformance]" srcIP="127.0.0.1:48856" hijacked=true
```

So bottom line (I think) is fixing the query string for proxy is breaking it for other things that rely on the query string to work, like exec.

---

With this PR as it stands currently, it has a safeguard in that it will not alter the query string if the location has a query string.  So in the case of exec, it will be safe.  And I believe in the case of proxy this also fixes the problem.  So I think this would be an improvement.  But I am open to other suggestions or ideas if someone else knows of a more comprehensive way to handle merging the query strings in a way that won't break exec.

**Comment by Jefftree** [comment]
Looks like k/k opened again for 1.22
/lgtm

**Comment by caesarxuchao** [comment]
I'm removing the triage/accepted label to let sig-auth take a look if using the query string from the request has any security concern.

/remove-triage accepted
/sig auth

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 2]
```diff
@@ -282,6 +282,11 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
...
+		// When a connection is upgraded, we want to keep the request querystring, even if the request location
+		// is not being used. However, if the location already has its own querystring, we should not overwrite
```
[the remark this answers is not in the document]
Comment (brianpursley) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:287]: I think it is a totally valid concern, and I would be OK with closing this PR and consider it a known limitation of proxy, that you just can't pass query string parameters on a proxied websocket request.  

Besides, this PR has been around over a year and hasn't really gotten a lot of interest.  Personally, I don't have a stake in whether it gets fixed or not, I just reproduced the issue and worked on a fix that I thought would work.

It certainly seems like an edge case I'll admit, and given that, I'm not sure it is worth the potential risk even though I don't immediately see a risk of adding a query string to the request.  As with all things security-related, you can be 99% right and 1% wrong and its the 1% that matters, so I get it.

I guess a workaround might be to [use `kubectl port-forward` instead](https://github.com/kubernetes/kubernetes/issues/89360#issuecomment-797207790)

[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:287]: it looks like for _non-upgrade_ requests, we use the query from the request unconditionally:

https://github.com/kubernetes/kubernetes/blob/14df10f60bbd68216f8ab9d3f30d1f938d7b659e/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L203-L204

to help decide whether this change is safe or needs to be limited to specific instances (like the node/pod/service proxy endpoints), can the following info be gathered?

1. enumerate the places we use `NewUpgradeAwareHandler` in the apiserver and kubelet
2. of those, determine which uses don't already honor the request query (by not setting `UseRequestLocation` to true, and by not explicitly honoring the request query like `staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go` does)
3. of those, determine which uses are targeting backends that could support upgrade requests
4. of those, determine which uses are not populating the backend query themselves 100% of the time
5. of those, determine which uses have known backend parameters that would be problematic to allow control over (the exec/attach endpoints especially come to mind, if the user could send a POST to the API server exec/attach endpoints with both a query and a body, get the body to be used for admission/audit inspection, then override the parameters with a query string)

[the remark this answers is not in the document]
Comment (brianpursley) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:287]: Thanks, @liggitt 

I've looked through the code and found the places where `NewUpgradeAwareHandler` is being used:
* pod exec/attach/portforward/proxy (`pkg/registry/core/pod/rest/subresource.go`)
* service proxy (`pkg/registry/core/service/proxy.go`)
* node proxy (`pkg/registry/core/node/rest/proxy.go`)
* kube-aggregator (`staging/src/k8s.io/kube-aggregator/pkg/apiserver`)
* kubectl proxy (`staging/src/k8s.io/kubectl/pkg/proxy`)

Of those, it appears that only kubectl proxy sets `UseRequestLocation` to true.

I'm not quite sure how to tell which of those are targeting backends that could support upgrade requests.  So that's where I left off.

---

However, I've been thinking it over and **want to get your thoughts on a different approach**:

What if instead of automatically applying the request query like this PR is doing, we add another option to `UpgradeAwareHandler`, called something like`UseRequestQuery`, which works in a way similar to `UseRequestLocation`?  

This option would make it an explicit choice to use the request query without the request location.

The advantage I see to this is that it would be possible to make this new behavior apply *only* to proxy, leaving exec, portforward, and attach to work the same as they do today.

I opened a draft PR to show an example of what I'm talking about. It still probably needs some work, but I created this to illustrate the idea:
https://github.com/kubernetes/kubernetes/pull/102002/files

[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:287]: requiring opt-in is certainly easier to reason about... I like the direction of that PR

would want to pick a good name, and think about / document well how it interacts with the other options on the proxier

[the remark this answers is not in the document]
Comment (brianpursley) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: I could go either way really.

Opt-in has the advantage of limiting the scope of the change, but it also seems to add some additional complexity (yet another flag).  While it is a safer change, it may not really be the best change from a code perspective due to the already somewhat large number of flags that affect how it works.

To add a point about the original approach of just always preserving the request query:

Doing that for upgraded requests would be consistent with what it does currently in the non-upgrade scenario:
https://github.com/kubernetes/kubernetes/blob/3d82e8e3c1eef729df36c56c97fc461c26a97738/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L203-L204

To simplify this PR as much as possible, I've updated my commit to just do exactly the same thing in the upgrade scenario as the non-upgrade scenario:
https://github.com/kubernetes/kubernetes/blob/3d82e8e3c1eef729df36c56c97fc461c26a97738/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L280-L281

This would bring into alignment how it handles the query in both the **upgrade** and **non-upgrade** scenarios so they behave the same in both cases.

Your previous points about how to determine whether this is safe in all cases are still valid and I will need to take some time to investigate further.
